src/shadowfax.py

```diff
@@ -201,17 +219,19 @@ async def shadowfax(message, guid, logger):
 
     paths = networkx.all_simple_paths(kg, source_node, target_node, 4)
     num_paths = 0
-    curie_path_mapping = defaultdict(list)
+    result_paths = []
     for path in paths:
         num_paths += 1
+        fits_constraint = False
         for curie in path:
             if curie not in [source_node, target_node] and curie in curie_info.keys():
-                # Handles constraints
-                if unpinned_node_category in curie_info[curie]["categories"]:
-                    curie_path_mapping[curie].append(path)
+                # Handles constraints, behavior may change depending decision handling multiple constraints
+                if intermediate_categories[0] in curie_info[curie]["categories"]:
+                    fits_constraint = True
+        if fits_constraint:
```

Comment:
What do/should we do if a path doesn't fit the constraint? Would we ever be getting back paths that don't fit?

Reply:
Yeah we'll get paths that don't fit the constraint. They just don't get returned as results and therefore the nodes and edges associated aren't added to the kg.

Reply:
Is it worth it to log when a path doesn't fit? Or are there a lot?

Reply:
No, I don't think that would make sense to log that a path doesn't fit. There can be a lot, it just depends on the query. Essentially, that just means that we generated a path that doesn't have "Gene" or whatever the intermediate category constraint it in the middle. The constraint is included as part of the query, so all of those extra paths would just clog up the logs with stuff that really is not relevant to the query.

Reply:
I'm willing to let this go, but I can't help but wonder why we're getting a bunch of intermediate paths that don't fit the constraint in the first place. And maybe I don't understand the nuances. But it seems like it would really help performance if we were able to query for the specific constraints instead of getting everything back and needing to filter afterwards.

Reply:
Yeah, unfortunately that isn't possible. We do our node selection first, and then find all the edges between the candidate nodes. After building the knowledge graph, then we can find the paths. Only after we find the paths can we determine if the path is lacking the intermediate category. It's not like every node has to fit the constraint, just one node in the path has to. There isn't a way to do it beforehand.